### ISSUE TITLE
mutating webhook were reported as uncompliant but manifest seems to be aligned with 1.22

### ISSUE THREAD

**Issue by Snehil03** [doitintl/kube-no-trouble#441]
Hi,

I am updating cluster so using kubent to validate .
It's the best utility tool , I know at the moment.

I am facing one issue that kubent refer below three as non-compliant.
```
kubent
9:51AM INF >>> Kube No Trouble `kubent` <<<
9:51AM INF version 0.7.0 (git sha d1bb4e5fd6550b533b2013671aa8419d923ee042)
9:51AM INF Initializing collectors and retrieving data
9:51AM INF Target K8s version is 1.22.16-eks-ffeb93d
9:51AM INF Retrieved 120 resources from collector name=Cluster
9:51AM WRN failed to discover supported resources for networking.k8s.io/v1beta1: the server could not find the requested resource
9:51AM INF Retrieved 482 resources from collector name="Helm v3"
9:51AM INF Loaded ruleset name=custom.rego.tmpl
9:51AM INF Loaded ruleset name=deprecated-1-16.rego
9:51AM INF Loaded ruleset name=deprecated-1-22.rego
9:51AM INF Loaded ruleset name=deprecated-1-25.rego
9:51AM INF Loaded ruleset name=deprecated-1-26.rego
9:51AM INF Loaded ruleset name=deprecated-future.rego
__________________________________________________________________________________________
>>> Deprecated APIs removed in 1.22 <<<
------------------------------------------------------------------------------------------
KIND                           NAMESPACE     NAME                     API_VERSION                            REPLACE_WITH (SINCE)
ClusterRoleBinding             <undefined>   eks:kube-proxy-windows   rbac.authorization.k8s.io/v1beta1      rbac.authorization.k8s.io/v1 (1.8.0)
Ingress                        <undefined>   elasticsearch-master     networking.k8s.io/v1beta1              networking.k8s.io/v1 (1.19.0)
MutatingWebhookConfiguration   <undefined>   pod-identity-webhook     admissionregistration.k8s.io/v1beta1   admissionregistration.k8s.io/v1 (1.16.0)
```
but while looking at the output of manifest .

I get below outcome :

```
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"admissionregistration.k8s.io/v1beta1","kind":"MutatingWebhookConfiguration","metadata":{"annotations":{},"name":"pod-identity-webhook"},"webhooks":[{"clientConfig":{"caBundle":,"url":"https://127.0.0.1:23443/mutate"},"name":"iam-for-pods.amazonaws.com","rules":[{"apiGroups":[""],"apiVersions":["v1"],"operations":["CREATE"],"resources":["pods"]}]}]}
  creationTimestamp: "2020-04-17T14:45:51Z"
  generation: 1
  name: pod-identity-webhook
  resourceVersion: "175"
  uid: 21ff8c1f-80ba-11ea-a190-0ac352e66ff0
webhooks:
- admissionReviewVersions:
  - v1beta1
  clientConfig:
    caBundle: 
    url: https://127.0.0.1:23443/mutate
  failurePolicy: Ignore
  matchPolicy: Exact
  name: iam-for-pods.amazonaws.com
  namespaceSelector: {}
  objectSelector: {}
  reinvocationPolicy: Never
  rules:
  - apiGroups:
    - ""
    apiVersions:
    - v1
    operations:
    - CREATE
    resources:
    - pods
    scope: '*'
  sideEffects: Unknown
  timeoutSeconds: 30
```

please let me know if these is the correct behaviour of kubent. I am using latest version of kubent .

**Comment by vojtechvelkjop** [comment]
Hi
same big issue for me because huge number of resource has not actual last-applied-configuration as you.
your apiVersion is correct
but
check your annotation
` kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"admissionregistration.k8s.io/v1beta1`
      
 so kubent thing you have bad API.
 question is why kubent checking annotation and not directly apiVersion of objects

**Comment by 0xErnie** [comment]
This error may be the result of the way that this tool fetches the objects from the API see
[kubepug#120](https://github.com/rikatz/kubepug/issues/120): Objects seem to be displayed through the older API-Version, even if they are already in the newest one.

**Comment by stepanstipl** [comment]
This is expected.

I think this is a bit of a misunderstanding of how K8S works internally regarding API versions. The `last-applied-configuration` annotation is how `kubent` is actually able to detect old versions (or one of the ways). You can't rely on the returned `apiVersion` from the live resource, as this is governed by the client, independently how the resource was created in the first place (that information is actually not stored anywhere, and internally the resource is stored in API version-independent format). I.e. you can get the same resource in all the supported API versions. I've explained a bit more about it in this comment https://github.com/doitintl/kube-no-trouble/issues/230#issuecomment-982717780 and in https://github.com/doitintl/kube-no-trouble/issues/276#issuecomment-1062928787.

TL;DR: If you have old `apiVersion` in the `last-applied-configuration`, it is likely the resource was created using this old API version. And as this would fail once you upgrade to the next minor K8s version, as the API won't be supported anymore, it is flagged by `kubent`.

I hope this makes sense. But let me know if you have any questions or if have misunderstood the issue. 👍

**Comment by robotica72** [comment]
I hate to start this question again, I understand the last-applied discussion - described very well and appreciated.  It's just one thing that still has me confused a bit:

I discovered a lab cluser that hasnt been upgraded in some time, its running 1.23.  I used kubent to look at the APIs as a test and it flagged a bunch of resources, but l'll ask about just one as an example:

Last applied for an Ingress resource says that it used networking.k8s.io/v1beta1 - I have the source manifest and the declared API is networking.k8s.io/v1   (and of course, you cant list v1beta1 in 1.23, so that lookup fails - the deployed resource is using v1, not v1beta1)  The Ingress resource is around 450 days old (as I said, discovered cluster that has been missed by a team - so I was asked to look)

The last applied doesnt have a date stamp, so I have to admit that Im not sure how this Ingress was deployed or updated when it was last run - but I have a few resources in this cluster with the same findings, source manifests are using the correct API but the last-applied shows an old API version.  This means that my reporting to my users would be flawed - assuming I understad what I am seeing and what has been mentioned on why last-applied was used?
